File: Problem-044/solution.js
function pentagonNumbers() {
    function pentagonal(n) {
        return n * (3 * n - 1) / 2;
    }

    function isPentagonal(num) {
        const x = 24 * num + 1;
        const root = Math.sqrt(x);

        return Number.isInteger(root);
    }

    const pentagonalNumbers = new Set();

    // Generate enough pentagonal numbers
    for (let n = 1; n <= 10000; n++) {
        pentagonalNumbers.add(pentagonal(n));
    }

    let smallestDifference = Infinity;

    for (let k = 2; k <= 10000; k++) {
        const pk = pentagonal(k);

        for (let j = k - 1; j >= 1; j--) {
            const pj = pentagonal(j);

            const difference = pk - pj;

            // We only need smaller differences
            if (difference >= smallestDifference) {
                break;
            }

            // Both sum and difference must be pentagonal
            if (
                pentagonalNumbers.has(difference) &&
                pentagonalNumbers.has(pk + pj)
            ) {
                smallestDifference = difference;
            }
        }
    }

    return smallestDifference;
}

File: Problem-044/solution.py
def pentagon_numbers():
    def pentagonal(n):
        return n * (3 * n - 1) // 2

    def is_pentagonal(num):
        if num <= 0:
            return False

        x = 24 * num + 1
        root = int(x ** 0.5)

        return root * root == x

    pentagonal_set = set()

    # Generate pentagonal numbers
    for n in range(1, 10000):
        pentagonal_set.add(pentagonal(n))

    smallest_difference = float("inf")

    for k in range(2, 10000):
        pk = pentagonal(k)

        for j in range(k - 1, 0, -1):
            pj = pentagonal(j)

            difference = pk - pj

            if difference >= smallest_difference:
                break

            total = pk + pj

            if difference in pentagonal_set and total in pentagonal_set:
                smallest_difference = difference

    return smallest_difference


print(pentagon_numbers())
